desc/grid.py

```diff
@@ -508,23 +579,133 @@ class Grid(_Grid):
         Spacing between nodes in each direction.
     weights : ndarray of float, size(num_nodes, )
         Quadrature weights for each node.
+    coordinates : str
+        Coordinates that are specified by the nodes.
+        raz : rho, alpha, zeta
+        rpz : rho, theta_PEST, zeta
+        rtz : rho, theta, zeta
+    period : tuple of float
+        Assumed periodicity for each coordinate.
+        Use np.inf to denote no periodicity.
+    source_grid : Grid
+        Grid from which coordinates were mapped from.
     sort : bool
         Whether to sort the nodes for use with FFT method.
+    is_meshgrid : bool
+        Whether this grid is separable into coordinate chunks.
+        Let the tuple (r, p, t) ∈ R³ denote a radial, poloidal, and toroidal
+        coordinate value. The is_meshgrid flag denotes whether any coordinate
+        can be iterated over along the relevant axis of the reshaped grid:
+        nodes.reshape(num_radial, num_poloidal, num_toroidal, 3).
     jitable : bool
         Whether to skip certain checks and conditionals that don't work under jit.
         Allows grid to be created on the fly with custom nodes, but weights, symmetry
-        etc may be wrong if grid contains duplicate nodes.
+        etc. may be wrong if grid contains duplicate nodes.
     """
 
+    @classmethod
+    def create_meshgrid(
+        cls,
+        nodes,
+        spacing=None,
+        coordinates="rtz",
+        period=(np.inf, 2 * np.pi, 2 * np.pi),
+        **kwargs,
+    ):
+        """Create a meshgrid from the given coordinates in a jitable manner.
+
+        Parameters
+        ----------
+        nodes : Array, Array, Array
+            Sorted unique values of each coordinate.
+        spacing : Array, Array, Array
+            Weights for integration. Defaults to a midpoint rule.
+        coordinates : str
+            Coordinates that are specified by the nodes a, b, c, respectively.
+            raz : rho, alpha, zeta
+            rpz : rho, theta_PEST, zeta
+            rtz : rho, theta, zeta
+        period : tuple of float
+            Assumed periodicity for each coordinate.
+            Use np.inf to denote no periodicity.
+
+        Returns
+        -------
+        grid : Grid
+            Meshgrid.
+
+        """
+        a, b, c = jnp.atleast_1d(*nodes)
+        if spacing is None:
+            errorif(coordinates[0] != "r", NotImplementedError)
+            da = _midpoint_spacing(a)
+            db = _periodic_spacing(b, period[1])[1]
+            dc = _periodic_spacing(c, period[2])[1]
+        else:
+            da, db, dc = spacing
+        nodes = jnp.column_stack(
+            list(map(jnp.ravel, jnp.meshgrid(a, b, c, indexing="ij")))
+        )
+        spacing = jnp.column_stack(
+            list(map(jnp.ravel, jnp.meshgrid(da, db, dc, indexing="ij")))
+        )
+        weights = spacing.prod(axis=1) if np.prod(period) == 4 * jnp.pi**2 else None
```

Comment:
can you explain this a bit?

Reply:
I should have written `period[1] * period[2] == 4 pi^2`.
I think there's two uses for this constructor.

1. jitably Make some tensor product grid over a surface using nodes for some coordinate system, (e.g. either `rho, theta, zeta`, or maybe `rho, theta_pest, zeta`, or `rho, alpha, zeta`). In this case period=(np.inf, 2pi, 2pi).
2. Make a grid following a few field-lines. E.g.  rho, alpha, zeta coordinates with zeta in [0, large number] and periwod=(np.inf, 2pi, np.inf). Here I don't assign quadrature weights for volume integration on  a field line following grid; eights would be nan in this case, since the spacing for the last node is (finite number % np.inf = nan).

Reply:
ok, i think it might be better to spell out that logic, or maybe just check the coordinates to determine if weights should be assigned. IE, if the grid is rtz over a single period, then `period[1] * period[2] == 4*np.pi**2/NFP` but it would still make sense to assign weights.

Also fwiw `x % np.inf == x if x > 0 else np.inf`